### ISSUE TITLE
Question: precedence of CLI options

### ISSUE THREAD

**Issue by aik099** [squizlabs/PHP_CodeSniffer#451]
When I specify CLI option in `ruleset.xml` and during run in command-line then what happens?

1. both values are used
2. command line value always wins
3. `ruleset.xml` value always wins

**Comment by VasekPurchart** [comment]
I have tested this for example with the `tab-width` option and the `ruleset.xml` value was the one that won.

I think for the sake of customization CLI values should be preffered, if you are heading this way :)

**Comment by gsherwood** [comment]
Because the ruleset is processed after the CLI is, it will override a value like tab-width and other config settings.

If the standard specifies that the tab-width must be x, then changing it on the command line is not a good idea. You'll potentially get unexpected results from the sniffs. The same goes for other config values as well.

**Comment by VasekPurchart** [comment]
It does not always have to be true and with some options, such as extensions it makes sense to enable the user to add his custom extension to what is defined in the standard (standard has only php, users wants to check php,phpt for example).

The current situation leads me to not using the configuration options in the standard at all.

**Comment by gsherwood** [comment]
> It does not always have to be true and with some options, such as extensions it makes sense to enable the user to add his custom extension to what is defined in the standard

I've just looked over all the options and extensions is the only one that makes sense to do a merge on, so I'll do that. If you can think of others that you would actually include on both the CLI and inside a ruleset, please let me know.

**Comment by gsherwood** [comment]
Forgot to the mention the other obvious one - exclude patterns. These are already merged.

**Comment by gsherwood** [comment]
Commit is here: https://github.com/squizlabs/PHP_CodeSniffer/commit/825f86a13d8c659061c3ce09d661640e0ee3f67f

**Comment by VasekPurchart** [comment]
I am not sure which options are supported (all?) in the XML `arg`, but from the options I look at in the CLI documentation it makes sense for me for almost all of them, maybe with the exception of the few ones which can mess with up actual standard such as `tab-width` or `encoding`.

The thing which confused me on the current behavior the most is the fact that you set some CLI parameters, you expect them to be used and they are not and you don't get any information on this fact - I think when user input is thrown in such a way, you should get an error/warning.

Because you don't even have to know that this change was introduced to the standard, for example I've just noticed that in 3e0cbf06fadf06e9bbaeaa1b5966538135f850f9 the tab-width was added to PSR2 ruleset, so that means that if someone was using the CLI options until that commit (as I am), they no longer need to and they probably shouldn't since they are not used any way.

Also I am not sure if merging is the right strategy, because again then there is no way for the user to override this configuration.

**Comment by gsherwood** [comment]
> but from the options I look at in the CLI documentation it makes sense for me for almost all of them

You can't merge most options. They are either on or off, so the ruleset takes over. Of course, you don't need to include these options inside a ruleset.

> and you don't get any information on this fact

You have to ask for information. You need to use the -vv command line argument to see how the ruleset is being parsed and what options are being applied.

**Comment by aik099** [comment]
I've looked which CLI options are used in my company with `phpcs` and `phpcbf` commands and was thinking to move some of them into `ruleset.xml` to make command, that people execute shorter.

Options I've added:
* tab-width
* encoding

Options I've not added:
* extensions

For me the only CLI options that makes sense to merge with ones, defined in `ruleset.xml` are:
* ignore
* extensions

**Comment by gsherwood** [comment]
> For me the only CLI options that makes sense to merge with ones, defined in ruleset.xml are:
> 
> ignore
> extensions

Those are the ones that are now being merged. Ignore was always merged, but extensions was being overwritten.

**Comment by VasekPurchart** [comment]
I am taking this from my experience with the end-users - e.g. the people who do not make the standards, or the integration with CI etc., but who just rarely need to change a few options.

> You can't merge most options. They are either on or off, so the ruleset takes over. Of course, you don't need to include these options inside a ruleset.

I think the merging shouldn't be done, because it might be confusing in the same way - you can not override this in any way if you are using a standard which is not under your control. Which makes them unnecessarily inflexible. The example I wrote with `extensions` can be used again here, only slightly modified - if the standard defines php,phpt, but the project does not want to check phpt (or maybe wants to run two different phpcs checks, one for each type - applications vs tests - and with different settings).

> You have to ask for information. You need to use the -vv command line argument to see how the ruleset is being parsed and what options are being applied.

This seems to me really odd from the user experience point of view. I think these verbose outputs should be used mainly for debugging. Situation where the user "gives direct order" and something completely different happens without a warning is imho not a good idea.

---

All of these situations would be solved, if the CLI parameters would always "win" if present.
I think most of tools also behave in this way, look for example at the git configuration - from the most wide to most specific, you can define:
* git defaults
* system wide config
* user wide config
* config for each repository
* and even then you can override this "defaults" with options on commands, because these are the most specific use-cases. This gives you the biggest freedom.

I see the parallel in PHPCS as:
* there are some default values from PHPCS
* standards can define default values which they think they work best with, so that user does not *have to* specify these manually
  * also if a standard is included in another one, the parent should take precedence (which I believe is the behavior currently implemented)
* user can always override any of these setting via the CLI options

From this discussion I see only this argument against this way:

> If the standard specifies that the tab-width must be x, then changing it on the command line is not a good idea. You'll potentially get unexpected results from the sniffs. The same goes for other config values as well.

Yes, that means if users have freedom to change the configuration values, they *can* shoot themselves in the leg. But for me this is always worth the flexibility. Because I don't think you can in any way solve all the use-cases, which the users can break.

---

If you really don't  want to support the behavior outlined in the previous section I would at least strongly recommend writing out the warnings if the configuration options are ignored.

This is similar to the situation when you type something like this:
``` bash
phpcs --foo .
```
and the program replies with
```
ERROR: option "--foo" not known.

Usage: ...
```

Yes, the program could run and ignore the invalid configuration option, but that would be really uncomfortable and for some programs even dangerous.

Ignoring the user given CLI options in PHPCS is for me the same thing. And this is not just some kind of "theoretical" debate, this comes from my experience with people/teams where I introduced PHPCS and they ran in this kind of confusion as well.

**Comment by gsherwood** [comment]
There is a clearly defined order of precedence. Use whatever level makes sense for the option. Nothing is mandatory. It is completely up to you.

1. Hard-code defaults (e., default standard is PEAR, report is "full")
2. User-defined config settings (--config-set --config-delete --config-show)
3. User-supplied CLI arguments
4. Standard-enforced CLI arguments

It should be fairly obvious that you shouldn't be enforcing many command line arguments in your standard. Some, like `tab-width` and `encoding` are sometimes mandatory. Others should only be included if you want to enforce a CLI option, or if you want to use ruleset.xml files as a set of pre-defined CLI argument groupings, such as like this:

ruleset.xml - just the sniffs, use this if you want to check files with the standard
default.xml - imports ruleset.xml and runs PHPCS using some pre-defined CLI args
ci.xml - imports ruleset.xml and runs PHPCS with CI-based arguments like where to write a checkstyle report

The current system provides a lot of flexibility for both developers who use PHPCS and developers who define custom coding standards. If CLI options are ignored, it is because a standard developer wants them ignored.

**Comment by VasekPurchart** [comment]
Yes, I understand how the current implementation works, what I am arguing is that switching points 3. and 4. would greatly improve flexibility.

The current situation would be like if Git had repository-wide configuration which could be only edited by a maintainer but not all the users - therefore enforcing certain ways of doing things.

I know you *can* create more files, where your preferences would go, but that is in my eyes really *uncomfortable* (I have never said that it is impossible).

In fact I am creating a custom standard, which would probably be used across many applications, maybe in time different teams and companies (much like the standards included in PHPCS). What I am worried about is that I have two options:

1. I will include the configuration options which I think suit the standard the best in the standard itself, but then I am limiting the usage for the end-user (he can no longer configure it via just adding CLI options), he *has* to create a file to override the defaults
2. I will not include the defaults directly in the standard, but since the are still some defaults, which should probably be used, I'll have to add them to (for example) a readme file stating that they should be used either via CLI or custom XML. Which again puts unnecessary requirements on the user.

So both options have significant disadvantages from my point of view and there is no other way I can see.

Also when you have a standard with default values and you just want to do some experimenting with the settings it is much easier to do this with CLI options rather than editing a file between your tries. Right now when I want to do this I have to delete all the options from the XMLs (my standard - if there were any and all the standards I include) and then I am finally free to use the CLI to it's full potential.

**Comment by gsherwood** [comment]
From version 3.0, the processing of ruleset options is a bit different. It now works by allowing the CLI options to override the ruleset options. From 2.5.0 and 3.0, the ruleset is also used even if files are specified on the command line. So you can use your normal ruleset (including a phpcs.xml file) and override parts of it using the command line.
